controllers/clusters/on_premises.go

```diff
@@ -93,302 +90,381 @@ func handleCreateOnPremisesClusterResources(ctx context.Context, b *onPremisesBo
-	gatewayDVName := fmt.Sprintf("%s-%s", models.GatewayDVPrefix, strings.ToLower(b.K8sObject.GetName()))
-	gatewayDV, err := createDV(
-		ctx,
-		b,
-		gatewayDVName,
-		b.ClusterStatus.DataCentres[0].ID,
-		gatewayDVSize,
-		true,
-	)
-	if err != nil {
-		return err
+	if newNode == nil ||
+		oldNode == nil {
+		return models.ErrNoNodeToReplace
 	}
 
-	gatewayCPU := resource.Quantity{}
-	gatewayCPU.Set(b.OnPremisesSpec.SSHGatewayCPU)
+	objectName := b.K8sObject.GetName()
+	vm := &virtcorev1.VirtualMachine{}
+	for i := 0; i < len(b.ClusterStatus.DataCentres[0].Nodes); i++ {
+		nodeName := fmt.Sprintf("%s-%d-%s", models.NodeVMPrefix, i, objectName)
+		err := b.K8sClient.Get(ctx, types.NamespacedName{
+			Namespace: b.K8sObject.GetNamespace(),
+			Name:      nodeName,
+		}, vm)
+		if client.IgnoreNotFound(err) != nil {
+			return err
+		}
+		if err == nil {
+			continue
+		}
 
-	gatewayMemory, err := resource.ParseQuantity(b.OnPremisesSpec.SSHGatewayMemory)
-	if err != nil {
-		return err
-	}
+		osDVName := fmt.Sprintf("%s-%d-%s", models.NodeOSDVPrefix, i, objectName)
+		osDV, err := reconcileDV(ctx, b, newNode.ID, osDVName, models.OSDVLabel)
+		if err != nil {
+			return err
+		}
 
-	gatewayName := fmt.Sprintf("%s-%s", models.GatewayVMPrefix, strings.ToLower(b.K8sObject.GetName()))
+		storageDVList := &cdiv1beta1.DataVolumeList{}
+		err = b.K8sClient.List(ctx, storageDVList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				models.NodeIDLabel: oldNode.ID,
+				models.DVRoleLabel: models.StorageDVLabel,
+			}),
+		})
+		if err != nil {
+			return err
+		}
 
-	gatewayVM := &virtcorev1.VirtualMachine{}
-	err = b.K8sClient.Get(ctx, types.NamespacedName{
-		Namespace: b.K8sObject.GetNamespace(),
-		Name:      gatewayName,
-	}, gatewayVM)
-	if client.IgnoreNotFound(err) != nil {
-		return err
-	}
-	if k8serrors.IsNotFound(err) {
-		gatewayVM, err = newVM(
-			ctx,
-			b,
-			gatewayName,
-			b.ClusterStatus.DataCentres[0].ID,
-			models.GatewayRack,
-			gatewayDV.Name,
-			gatewayCPU,
-			gatewayMemory)
+		storageDV := &cdiv1beta1.DataVolume{}
+		storageDVName := fmt.Sprintf("%s-%d-%s", models.NodeDVPrefix, i, objectName)
+		err = b.K8sClient.Get(ctx, types.NamespacedName{
+			Namespace: b.K8sObject.GetNamespace(),
+			Name:      storageDVName,
+		}, storageDV)
 		if err != nil {
 			return err
 		}
-		err = b.K8sClient.Create(ctx, gatewayVM)
+
+		patch := client.MergeFrom(storageDV.DeepCopy())
+		storageDV.Labels[models.NodeIDLabel] = newNode.ID
+		err = b.K8sClient.Patch(ctx, storageDV, patch)
 		if err != nil {
 			return err
 		}
-	}
 
-	gatewaySvcName := fmt.Sprintf("%s-%s", models.GatewaySvcPrefix, gatewayName)
-	gatewayExposeService := &k8scorev1.Service{}
-	err = b.K8sClient.Get(ctx, types.NamespacedName{
-		Namespace: b.K8sObject.GetNamespace(),
-		Name:      gatewaySvcName,
-	}, gatewayExposeService)
+		svc := &k8scorev1.Service{}
+		svcName := fmt.Sprintf("%s-%s", models.ClusterIPServiceRoleLabel, nodeName)
```

Comment:
Why do we set a new label for a ClusterIP service only? Should we set updated labels for the LB service if the cluster is a public one?